// @carbon/charts/dist/angle-utils-CQDEbkP4.mjs
import { P as Zt, Q as De, R as Xt, S as Oe, U as Qt, V as Jt, W as ut, X as Kt, Y as We, Z as en, _ as tn, g as ie, p as $e } from "./color-scale-utils-ycwO-Ibq.mjs";
import { $ as pe, o as B, T as V } from "./axis-scales-CPuwbFQf.mjs";
var nn = /\s/;
function rn(e) {
  for (var t = e.length; t-- && nn.test(e.charAt(t)); )
    ;
  return t;
}
var on = /^\s+/;
function sn(e) {
  return e && e.slice(0, rn(e) + 1).replace(on, "");
}
var Ye = NaN, an = /^[-+]0x[0-9a-f]+$/i, un = /^0b[01]+$/i, cn = /^0o[0-7]+$/i, ln = parseInt;
function _e(e) {
  if (typeof e == "number")
    return e;
  if (Zt(e))
    return Ye;
  if (De(e)) {
    var t = typeof e.valueOf == "function" ? e.valueOf() : e;
    e = De(t) ? t + "" : t;
  }
  if (typeof e != "string")
    return e === 0 ? e : +e;
  e = sn(e);
  var n = un.test(e);
  return n || cn.test(e) ? ln(e.slice(2), n ? 2 : 8) : an.test(e) ? Ye : +e;
}
var we = function() {
  return Xt.Date.now();
}, fn = "Expected a function", hn = Math.max, gn = Math.min;
function mn(e, t, n) {
  var r, o, i, s, a, c, h = 0, m = !1, d = !1, g = !0;
  if (typeof e != "function")
    throw new TypeError(fn);
  t = _e(t) || 0, De(n) && (m = !!n.leading, d = "maxWait" in n, i = d ? hn(_e(n.maxWait) || 0, t) : i, g = "trailing" in n ? !!n.trailing : g);
  function T(f) {
    var C = r, O = o;
    return r = o = void 0, h = f, s = e.apply(O, C), s;
  }
  function M(f) {
    return h = f, a = setTimeout(k, t), m ? T(f) : s;
  }
  function U(f) {
    var C = f - c, O = f - h, ce = t - C;
    return d ? gn(ce, i - O) : ce;
  }
  function I(f) {
    var C = f - c, O = f - h;
    return c === void 0 || C >= t || C < 0 || d && O >= i;
  }
  function k() {
    var f = we();
    if (I(f))
      return Q(f);
    a = setTimeout(k, U(f));
  }
  function Q(f) {
    return a = void 0, g && r ? T(f) : (r = o = void 0, s);
  }
  function A() {
    a !== void 0 && clearTimeout(a), h = 0, r = c = o = a = void 0;
  }
  function P() {
    return a === void 0 ? s : Q(we());
  }
  function G() {
    var f = we(), C = I(f);
    if (r = arguments, o = this, c = f, C) {
      if (a === void 0)
        return M(c);
      if (d)
        return clearTimeout(a), a = setTimeout(k, t), T(c);
    }
    return a === void 0 && (a = setTimeout(k, t)), s;
  }
  return G.cancel = A, G.flush = P, G;
}
function dn(e, t, n) {
  e = +e, t = +t, n = (o = arguments.length) < 2 ? (t = e, e = 0, 1) : o < 3 ? 1 : +n;
  for (var r = -1, o = Math.max(0, Math.ceil((t - e) / n)) | 0, i = new Array(o); ++r < o; )
    i[r] = e + r * n;
  return i;
}
function D(e) {
  return typeof e == "string" ? new Oe([[document.querySelector(e)]], [document.documentElement]) : new Oe([[e]], Qt);
}
function yn() {
  var e = Jt().unknown(void 0), t = e.domain, n = e.range, r = 0, o = 1, i, s, a = !1, c = 0, h = 0, m = 0.5;
  delete e.unknown;
  function d() {
    var g = t().length, T = o < r, M = T ? o : r, U = T ? r : o;
    i = (U - M) / Math.max(1, g - c + h * 2), a && (i = Math.floor(i)), M += (U - M - i * (g - c)) * m, s = i * (1 - c), a && (M = Math.round(M), s = Math.round(s));
    var I = dn(g).map(function(k) {
      return M + i * k;
    });
    return n(T ? I.reverse() : I);
  }
  return e.domain = function(g) {
    return arguments.length ? (t(g), d()) : t();
  }, e.range = function(g) {
    return arguments.length ? ([r, o] = g, r = +r, o = +o, d()) : [r, o];
  }, e.rangeRound = function(g) {
    return [r, o] = g, r = +r, o = +o, a = !0, d();
  }, e.bandwidth = function() {
    return s;
  }, e.step = function() {
    return i;
  }, e.round = function(g) {
    return arguments.length ? (a = !!g, d()) : a;
  }, e.padding = function(g) {
    return arguments.length ? (c = Math.min(1, h = +g), d()) : c;
  }, e.paddingInner = function(g) {
    return arguments.length ? (c = Math.min(1, g), d()) : c;
  }, e.paddingOuter = function(g) {
    return arguments.length ? (h = +g, d()) : h;
  }, e.align = function(g) {
    return arguments.length ? (m = Math.max(0, Math.min(1, g)), d()) : m;
  }, e.copy = function() {
    return yn(t(), [r, o]).round(a).paddingInner(c).paddingOuter(h).align(m);
  }, ut.apply(d(), arguments);
}
function pn(e, t) {
  e = e.slice();
  var n = 0, r = e.length - 1, o = e[n], i = e[r], s;
  return i < o && (s = n, n = r, r = s, s = o, o = i, i = s), e[n] = t.floor(o), e[r] = t.ceil(i), e;
}
const Te = /* @__PURE__ */ new Date(), Se = /* @__PURE__ */ new Date();
function x(e, t, n, r) {
  function o(i) {
    return e(i = arguments.length === 0 ? /* @__PURE__ */ new Date() : /* @__PURE__ */ new Date(+i)), i;
  }
  return o.floor = (i) => (e(i = /* @__PURE__ */ new Date(+i)), i), o.ceil = (i) => (e(i = new Date(i - 1)), t(i, 1), e(i), i), o.round = (i) => {
    const s = o(i), a = o.ceil(i);
    return i - s < a - i ? s : a;
  }, o.offset = (i, s) => (t(i = /* @__PURE__ */ new Date(+i), s == null ? 1 : Math.floor(s)), i), o.range = (i, s, a) => {
    const c = [];
    if (i = o.ceil(i), a = a == null ? 1 : Math.floor(a), !(i < s) || !(a > 0)) return c;
    let h;
    do
      c.push(h = /* @__PURE__ */ new Date(+i)), t(i, a), e(i);
    while (h < i && i < s);
    return c;
  }, o.filter = (i) => x((s) => {
    if (s >= s) for (; e(s), !i(s); ) s.setTime(s - 1);
  }, (s, a) => {
    if (s >= s)
      if (a < 0) for (; ++a <= 0; )
        for (; t(s, -1), !i(s); )
          ;
      else for (; --a >= 0; )
        for (; t(s, 1), !i(s); )
          ;
  }), n && (o.count = (i, s) => (Te.setTime(+i), Se.setTime(+s), e(Te), e(Se), Math.floor(n(Te, Se))), o.every = (i) => (i = Math.floor(i), !isFinite(i) || !(i > 0) ? null : i > 1 ? o.filter(r ? (s) => r(s) % i === 0 : (s) => o.count(0, s) % i === 0) : o)), o;
}
const le = x(() => {
}, (e, t) => {
  e.setTime(+e + t);
}, (e, t) => t - e);
le.every = (e) => (e = Math.floor(e), !isFinite(e) || !(e > 0) ? null : e > 1 ? x((t) => {
  t.setTime(Math.floor(t / e) * e);
}, (t, n) => {
  t.setTime(+t + n * e);
}, (t, n) => (n - t) / e) : le);
le.range;
const W = 1e3, H = W * 60, $ = H * 60, Y = $ * 24, Ue = Y * 7, Be = Y * 30, Ce = Y * 365, te = x((e) => {
  e.setTime(e - e.getMilliseconds());
}, (e, t) => {
  e.setTime(+e + t * W);
}, (e, t) => (t - e) / W, (e) => e.getUTCSeconds());
te.range;
const Fe = x((e) => {
  e.setTime(e - e.getMilliseconds() - e.getSeconds() * W);
}, (e, t) => {
  e.setTime(+e + t * H);
}, (e, t) => (t - e) / H, (e) => e.getMinutes());
Fe.range;
const wn = x((e) => {
  e.setUTCSeconds(0, 0);
}, (e, t) => {
  e.setTime(+e + t * H);
}, (e, t) => (t - e) / H, (e) => e.getUTCMinutes());
wn.range;
const Re = x((e) => {
  e.setTime(e - e.getMilliseconds() - e.getSeconds() * W - e.getMinutes() * H);
}, (e, t) => {
  e.setTime(+e + t * $);
}, (e, t) => (t - e) / $, (e) => e.getHours());
Re.range;
const Tn = x((e) => {
  e.setUTCMinutes(0, 0, 0);
}, (e, t) => {
  e.setTime(+e + t * $);
}, (e, t) => (t - e) / $, (e) => e.getUTCHours());
Tn.range;
const ue = x(
  (e) => e.setHours(0, 0, 0, 0),
  (e, t) => e.setDate(e.getDate() + t),
  (e, t) => (t - e - (t.getTimezoneOffset() - e.getTimezoneOffset()) * H) / Y,
  (e) => e.getDate() - 1
);
ue.range;
const Le = x((e) => {
  e.setUTCHours(0, 0, 0, 0);
}, (e, t) => {
  e.setUTCDate(e.getUTCDate() + t);
}, (e, t) => (t - e) / Y, (e) => e.getUTCDate() - 1);
Le.range;
const Sn = x((e) => {
  e.setUTCHours(0, 0, 0, 0);
}, (e, t) => {
  e.setUTCDate(e.getUTCDate() + t);
}, (e, t) => (t - e) / Y, (e) => Math.floor(e / Y));
Sn.range;
function Z(e) {
  return x((t) => {
    t.setDate(t.getDate() - (t.getDay() + 7 - e) % 7), t.setHours(0, 0, 0, 0);
  }, (t, n) => {
    t.setDate(t.getDate() + n * 7);
  }, (t, n) => (n - t - (n.getTimezoneOffset() - t.getTimezoneOffset()) * H) / Ue);
}
const de = Z(0), fe = Z(1), Cn = Z(2), Mn = Z(3), ne = Z(4), xn = Z(5), bn = Z(6);
de.range;
fe.range;
Cn.range;
Mn.range;
ne.range;
xn.range;
bn.range;
function X(e) {
  return x((t) => {
    t.setUTCDate(t.getUTCDate() - (t.getUTCDay() + 7 - e) % 7), t.setUTCHours(0, 0, 0, 0);
  }, (t, n) => {
    t.setUTCDate(t.getUTCDate() + n * 7);
  }, (t, n) => (n - t) / Ue);
}
const ct = X(0), he = X(1), Dn = X(2), En = X(3), re = X(4), vn = X(5), Un = X(6);
ct.range;
he.range;
Dn.range;
En.range;
re.range;
vn.range;
Un.range;
const He = x((e) => {
  e.setDate(1), e.setHours(0, 0, 0, 0);
}, (e, t) => {
  e.setMonth(e.getMonth() + t);
}, (e, t) => t.getMonth() - e.getMonth() + (t.getFullYear() - e.getFullYear()) * 12, (e) => e.getMonth());
He.range;
const Fn = x((e) => {
  e.setUTCDate(1), e.setUTCHours(0, 0, 0, 0);
}, (e, t) => {
  e.setUTCMonth(e.getUTCMonth() + t);
}, (e, t) => t.getUTCMonth() - e.getUTCMonth() + (t.getUTCFullYear() - e.getUTCFullYear()) * 12, (e) => e.getUTCMonth());
Fn.range;
const _ = x((e) => {
  e.setMonth(0, 1), e.setHours(0, 0, 0, 0);
}, (e, t) => {
  e.setFullYear(e.getFullYear() + t);
}, (e, t) => t.getFullYear() - e.getFullYear(), (e) => e.getFullYear());
_.every = (e) => !isFinite(e = Math.floor(e)) || !(e > 0) ? null : x((t) => {
  t.setFullYear(Math.floor(t.getFullYear() / e) * e), t.setMonth(0, 1), t.setHours(0, 0, 0, 0);
}, (t, n) => {
  t.setFullYear(t.getFullYear() + n * e);
});
_.range;
const j = x((e) => {
  e.setUTCMonth(0, 1), e.setUTCHours(0, 0, 0, 0);
}, (e, t) => {
  e.setUTCFullYear(e.getUTCFullYear() + t);
}, (e, t) => t.getUTCFullYear() - e.getUTCFullYear(), (e) => e.getUTCFullYear());
j.every = (e) => !isFinite(e = Math.floor(e)) || !(e > 0) ? null : x((t) => {
  t.setUTCFullYear(Math.floor(t.getUTCFullYear() / e) * e), t.setUTCMonth(0, 1), t.setUTCHours(0, 0, 0, 0);
}, (t, n) => {
  t.setUTCFullYear(t.getUTCFullYear() + n * e);
});
j.range;
function Rn(e, t, n, r, o, i) {
  const s = [
    [te, 1, W],
    [te, 5, 5 * W],
    [te, 15, 15 * W],
    [te, 30, 30 * W],
    [i, 1, H],
    [i, 5, 5 * H],
    [i, 15, 15 * H],
    [i, 30, 30 * H],
    [o, 1, $],
    [o, 3, 3 * $],
    [o, 6, 6 * $],
    [o, 12, 12 * $],
    [r, 1, Y],
    [r, 2, 2 * Y],
    [n, 1, Ue],
    [t, 1, Be],
    [t, 3, 3 * Be],
    [e, 1, Ce]
  ];
  function a(h, m, d) {
    const g = m < h;
    g && ([h, m] = [m, h]);
    const T = d && typeof d.range == "function" ? d : c(h, m, d), M = T ? T.range(h, +m + 1) : [];
    return g ? M.reverse() : M;
  }
  function c(h, m, d) {
    const g = Math.abs(m - h) / d, T = Kt(([, , I]) => I).right(s, g);
    if (T === s.length) return e.every(We(h / Ce, m / Ce, d));
    if (T === 0) return le.every(Math.max(We(h, m, d), 1));
    const [M, U] = s[g / s[T - 1][2] < s[T][2] / g ? T - 1 : T];
    return M.every(U);
  }
  return [a, c];
}
const [Ln, Hn] = Rn(_, He, de, ue, Re, Fe);
function Me(e) {
  if (0 <= e.y && e.y < 100) {
    var t = new Date(-1, e.m, e.d, e.H, e.M, e.S, e.L);
    return t.setFullYear(e.y), t;
  }
  return new Date(e.y, e.m, e.d, e.H, e.M, e.S, e.L);
}
function xe(e) {
  if (0 <= e.y && e.y < 100) {
    var t = new Date(Date.UTC(-1, e.m, e.d, e.H, e.M, e.S, e.L));
    return t.setUTCFullYear(e.y), t;
  }
  return new Date(Date.UTC(e.y, e.m, e.d, e.H, e.M, e.S, e.L));
}
function oe(e, t, n) {
  return { y: e, m: t, d: n, H: 0, M: 0, S: 0, L: 0 };
}
function In(e) {
  var t = e.dateTime, n = e.date, r = e.time, o = e.periods, i = e.days, s = e.shortDays, a = e.months, c = e.shortMonths, h = se(o), m = ae(o), d = se(i), g = ae(i), T = se(s), M = ae(s), U = se(a), I = ae(a), k = se(c), Q = ae(c), A = {
    a: Ot,
    A: Wt,
    b: $t,
    B: Yt,
    c: null,
    d: je,
    e: je,
    f: nr,
    g: hr,
    G: mr,
    H: Kn,
    I: er,
    j: tr,
    L: lt,
    m: rr,
    M: ir,
    p: _t,
    q: Bt,
    Q: Qe,
    s: Je,
    S: or,
    u: sr,
    U: ar,
    V: ur,
    w: cr,
    W: lr,
    x: null,
    X: null,
    y: fr,
    Y: gr,
    Z: dr,
    "%": Xe
  }, P = {
    a: Vt,
    A: zt,
    b: qt,
    B: Gt,
    c: null,
    d: Ze,
    e: Ze,
    f: Tr,
    g: Fr,
    G: Lr,
    H: yr,
    I: pr,
    j: wr,
    L: ht,
    m: Sr,
    M: Cr,
    p: Nt,
    q: jt,
    Q: Qe,
    s: Je,
    S: Mr,
    u: xr,
    U: br,
    V: Dr,
    w: Er,
    W: vr,
    x: null,
    X: null,
    y: Ur,
    Y: Rr,
    Z: Hr,
    "%": Xe
  }, G = {
    a: Rt,
    A: Lt,
    b: Ht,
    B: It,
    c: At,
    d: Ge,
    e: Ge,
    f: Zn,
    g: qe,
    G: ze,
    H: Ne,
    I: Ne,
    j: qn,
    L: jn,
    m: zn,
    M: Gn,
    p: ce,
    q: Vn,
    Q: Qn,
    s: Jn,
    S: Nn,
    u: Wn,
    U: $n,
    V: Yn,
    w: On,
    W: _n,
    x: Pt,
    X: kt,
    y: qe,
    Y: ze,
    Z: Bn,
    "%": Xn
  };
  A.x = f(n, A), A.X = f(r, A), A.c = f(t, A), P.x = f(n, P), P.X = f(r, P), P.c = f(t, P);
  function f(l, y) {
    return function(p) {
      var u = [], E = -1, S = 0, F = l.length, R, N, ke;
      for (p instanceof Date || (p = /* @__PURE__ */ new Date(+p)); ++E < F; )
        l.charCodeAt(E) === 37 && (u.push(l.slice(S, E)), (N = Ve[R = l.charAt(++E)]) != null ? R = l.charAt(++E) : N = R === "e" ? " " : "0", (ke = y[R]) && (R = ke(p, N)), u.push(R), S = E + 1);
      return u.push(l.slice(S, E)), u.join("");
    };
  }
  function C(l, y) {
    return function(p) {
      var u = oe(1900, void 0, 1), E = O(u, l, p += "", 0), S, F;
      if (E != p.length) return null;
      if ("Q" in u) return new Date(u.Q);
      if ("s" in u) return new Date(u.s * 1e3 + ("L" in u ? u.L : 0));
      if (y && !("Z" in u) && (u.Z = 0), "p" in u && (u.H = u.H % 12 + u.p * 12), u.m === void 0 && (u.m = "q" in u ? u.q : 0), "V" in u) {
        if (u.V < 1 || u.V > 53) return null;
        "w" in u || (u.w = 1), "Z" in u ? (S = xe(oe(u.y, 0, 1)), F = S.getUTCDay(), S = F > 4 || F === 0 ? he.ceil(S) : he(S), S = Le.offset(S, (u.V - 1) * 7), u.y = S.getUTCFullYear(), u.m = S.getUTCMonth(), u.d = S.getUTCDate() + (u.w + 6) % 7) : (S = Me(oe(u.y, 0, 1)), F = S.getDay(), S = F > 4 || F === 0 ? fe.ceil(S) : fe(S), S = ue.offset(S, (u.V - 1) * 7), u.y = S.getFullYear(), u.m = S.getMonth(), u.d = S.getDate() + (u.w + 6) % 7);
      } else ("W" in u || "U" in u) && ("w" in u || (u.w = "u" in u ? u.u % 7 : "W" in u ? 1 : 0), F = "Z" in u ? xe(oe(u.y, 0, 1)).getUTCDay() : Me(oe(u.y, 0, 1)).getDay(), u.m = 0, u.d = "W" in u ? (u.w + 6) % 7 + u.W * 7 - (F + 5) % 7 : u.w + u.U * 7 - (F + 6) % 7);
      return "Z" in u ? (u.H += u.Z / 100 | 0, u.M += u.Z % 100, xe(u)) : Me(u);
    };
  }
  function O(l, y, p, u) {
    for (var E = 0, S = y.length, F = p.length, R, N; E < S; ) {
      if (u >= F) return -1;
      if (R = y.charCodeAt(E++), R === 37) {
        if (R = y.charAt(E++), N = G[R in Ve ? y.charAt(E++) : R], !N || (u = N(l, p, u)) < 0) return -1;
      } else if (R != p.charCodeAt(u++))
        return -1;
    }
    return u;
  }
  function ce(l, y, p) {
    var u = h.exec(y.slice(p));
    return u ? (l.p = m.get(u[0].toLowerCase()), p + u[0].length) : -1;
  }
  function Rt(l, y, p) {
    var u = T.exec(y.slice(p));
    return u ? (l.w = M.get(u[0].toLowerCase()), p + u[0].length) : -1;
  }
  function Lt(l, y, p) {
    var u = d.exec(y.slice(p));
    return u ? (l.w = g.get(u[0].toLowerCase()), p + u[0].length) : -1;
  }
  function Ht(l, y, p) {
    var u = k.exec(y.slice(p));
    return u ? (l.m = Q.get(u[0].toLowerCase()), p + u[0].length) : -1;
  }
  function It(l, y, p) {
    var u = U.exec(y.slice(p));
    return u ? (l.m = I.get(u[0].toLowerCase()), p + u[0].length) : -1;
  }
  function At(l, y, p) {
    return O(l, t, y, p);
  }
  function Pt(l, y, p) {
    return O(l, n, y, p);
  }
  function kt(l, y, p) {
    return O(l, r, y, p);
  }
  function Ot(l) {
    return s[l.getDay()];
  }
  function Wt(l) {
    return i[l.getDay()];
  }
  function $t(l) {
    return c[l.getMonth()];
  }
  function Yt(l) {
    return a[l.getMonth()];
  }
  function _t(l) {
    return o[+(l.getHours() >= 12)];
  }
  function Bt(l) {
    return 1 + ~~(l.getMonth() / 3);
  }
  function Vt(l) {
    return s[l.getUTCDay()];
  }
  function zt(l) {
    return i[l.getUTCDay()];
  }
  function qt(l) {
    return c[l.getUTCMonth()];
  }
  function Gt(l) {
    return a[l.getUTCMonth()];
  }
  function Nt(l) {
    return o[+(l.getUTCHours() >= 12)];
  }
  function jt(l) {
    return 1 + ~~(l.getUTCMonth() / 3);
  }
  return {
    format: function(l) {
      var y = f(l += "", A);
      return y.toString = function() {
        return l;
      }, y;
    },
    parse: function(l) {
      var y = C(l += "", !1);
      return y.toString = function() {
        return l;
      }, y;
    },
    utcFormat: function(l) {
      var y = f(l += "", P);
      return y.toString = function() {
        return l;
      }, y;
    },
    utcParse: function(l) {
      var y = C(l += "", !0);
      return y.toString = function() {
        return l;
      }, y;
    }
  };
}
var Ve = { "-": "", _: " ", 0: "0" }, b = /^\s*\d+/, An = /^%/, Pn = /[\\^$*+?|[\]().{}]/g;
function w(e, t, n) {
  var r = e < 0 ? "-" : "", o = (r ? -e : e) + "", i = o.length;
  return r + (i < n ? new Array(n - i + 1).join(t) + o : o);
}
function kn(e) {
  return e.replace(Pn, "\\$&");
}
function se(e) {
  return new RegExp("^(?:" + e.map(kn).join("|") + ")", "i");
}
function ae(e) {
  return new Map(e.map((t, n) => [t.toLowerCase(), n]));
}
function On(e, t, n) {
  var r = b.exec(t.slice(n, n + 1));
  return r ? (e.w = +r[0], n + r[0].length) : -1;
}
function Wn(e, t, n) {
  var r = b.exec(t.slice(n, n + 1));
  return r ? (e.u = +r[0], n + r[0].length) : -1;
}
function $n(e, t, n) {
  var r = b.exec(t.slice(n, n + 2));
  return r ? (e.U = +r[0], n + r[0].length) : -1;
}
function Yn(e, t, n) {
  var r = b.exec(t.slice(n, n + 2));
  return r ? (e.V = +r[0], n + r[0].length) : -1;
}
function _n(e, t, n) {
  var r = b.exec(t.slice(n, n + 2));
  return r ? (e.W = +r[0], n + r[0].length) : -1;
}
function ze(e, t, n) {
  var r = b.exec(t.slice(n, n + 4));
  return r ? (e.y = +r[0], n + r[0].length) : -1;
}
function qe(e, t, n) {
  var r = b.exec(t.slice(n, n + 2));
  return r ? (e.y = +r[0] + (+r[0] > 68 ? 1900 : 2e3), n + r[0].length) : -1;
}
function Bn(e, t, n) {
  var r = /^(Z)|([+-]\d\d)(?::?(\d\d))?/.exec(t.slice(n, n + 6));
  return r ? (e.Z = r[1] ? 0 : -(r[2] + (r[3] || "00")), n + r[0].length) : -1;
}
function Vn(e, t, n) {
  var r = b.exec(t.slice(n, n + 1));
  return r ? (e.q = r[0] * 3 - 3, n + r[0].length) : -1;
}
function zn(e, t, n) {
  var r = b.exec(t.slice(n, n + 2));
  return r ? (e.m = r[0] - 1, n + r[0].length) : -1;
}
function Ge(e, t, n) {
  var r = b.exec(t.slice(n, n + 2));
  return r ? (e.d = +r[0], n + r[0].length) : -1;
}
function qn(e, t, n) {
  var r = b.exec(t.slice(n, n + 3));
  return r ? (e.m = 0, e.d = +r[0], n + r[0].length) : -1;
}
function Ne(e, t, n) {
  var r = b.exec(t.slice(n, n + 2));
  return r ? (e.H = +r[0], n + r[0].length) : -1;
}
function Gn(e, t, n) {
  var r = b.exec(t.slice(n, n + 2));
  return r ? (e.M = +r[0], n + r[0].length) : -1;
}
function Nn(e, t, n) {
  var r = b.exec(t.slice(n, n + 2));
  return r ? (e.S = +r[0], n + r[0].length) : -1;
}
function jn(e, t, n) {
  var r = b.exec(t.slice(n, n + 3));
  return r ? (e.L = +r[0], n + r[0].length) : -1;
}
function Zn(e, t, n) {
  var r = b.exec(t.slice(n, n + 6));
  return r ? (e.L = Math.floor(r[0] / 1e3), n + r[0].length) : -1;
}
function Xn(e, t, n) {
  var r = An.exec(t.slice(n, n + 1));
  return r ? n + r[0].length : -1;
}
function Qn(e, t, n) {
  var r = b.exec(t.slice(n));
  return r ? (e.Q = +r[0], n + r[0].length) : -1;
}
function Jn(e, t, n) {
  var r = b.exec(t.slice(n));
  return r ? (e.s = +r[0], n + r[0].length) : -1;
}
function je(e, t) {
  return w(e.getDate(), t, 2);
}
function Kn(e, t) {
  return w(e.getHours(), t, 2);
}
function er(e, t) {
  return w(e.getHours() % 12 || 12, t, 2);
}
function tr(e, t) {
  return w(1 + ue.count(_(e), e), t, 3);
}
function lt(e, t) {
  return w(e.getMilliseconds(), t, 3);
}
function nr(e, t) {
  return lt(e, t) + "000";
}
function rr(e, t) {
  return w(e.getMonth() + 1, t, 2);
}
function ir(e, t) {
  return w(e.getMinutes(), t, 2);
}
function or(e, t) {
  return w(e.getSeconds(), t, 2);
}
function sr(e) {
  var t = e.getDay();
  return t === 0 ? 7 : t;
}
function ar(e, t) {
  return w(de.count(_(e) - 1, e), t, 2);
}
function ft(e) {
  var t = e.getDay();
  return t >= 4 || t === 0 ? ne(e) : ne.ceil(e);
}
function ur(e, t) {
  return e = ft(e), w(ne.count(_(e), e) + (_(e).getDay() === 4), t, 2);
}
function cr(e) {
  return e.getDay();
}
function lr(e, t) {
  return w(fe.count(_(e) - 1, e), t, 2);
}
function fr(e, t) {
  return w(e.getFullYear() % 100, t, 2);
}
function hr(e, t) {
  return e = ft(e), w(e.getFullYear() % 100, t, 2);
}
function gr(e, t) {
  return w(e.getFullYear() % 1e4, t, 4);
}
function mr(e, t) {
  var n = e.getDay();
  return e = n >= 4 || n === 0 ? ne(e) : ne.ceil(e), w(e.getFullYear() % 1e4, t, 4);
}
function dr(e) {
  var t = e.getTimezoneOffset();
  return (t > 0 ? "-" : (t *= -1, "+")) + w(t / 60 | 0, "0", 2) + w(t % 60, "0", 2);
}
function Ze(e, t) {
  return w(e.getUTCDate(), t, 2);
}
function yr(e, t) {
  return w(e.getUTCHours(), t, 2);
}
function pr(e, t) {
  return w(e.getUTCHours() % 12 || 12, t, 2);
}
function wr(e, t) {
  return w(1 + Le.count(j(e), e), t, 3);
}
function ht(e, t) {
  return w(e.getUTCMilliseconds(), t, 3);
}
function Tr(e, t) {
  return ht(e, t) + "000";
}
function Sr(e, t) {
  return w(e.getUTCMonth() + 1, t, 2);
}
function Cr(e, t) {
  return w(e.getUTCMinutes(), t, 2);
}
function Mr(e, t) {
  return w(e.getUTCSeconds(), t, 2);
}
function xr(e) {
  var t = e.getUTCDay();
  return t === 0 ? 7 : t;
}
function br(e, t) {
  return w(ct.count(j(e) - 1, e), t, 2);
}
function gt(e) {
  var t = e.getUTCDay();
  return t >= 4 || t === 0 ? re(e) : re.ceil(e);
}
function Dr(e, t) {
  return e = gt(e), w(re.count(j(e), e) + (j(e).getUTCDay() === 4), t, 2);
}
function Er(e) {
  return e.getUTCDay();
}
function vr(e, t) {
  return w(he.count(j(e) - 1, e), t, 2);
}
function Ur(e, t) {
  return w(e.getUTCFullYear() % 100, t, 2);
}
function Fr(e, t) {
  return e = gt(e), w(e.getUTCFullYear() % 100, t, 2);
}
function Rr(e, t) {
  return w(e.getUTCFullYear() % 1e4, t, 4);
}
function Lr(e, t) {
  var n = e.getUTCDay();
  return e = n >= 4 || n === 0 ? re(e) : re.ceil(e), w(e.getUTCFullYear() % 1e4, t, 4);
}
function Hr() {
  return "+0000";
}
function Xe() {
  return "%";
}
function Qe(e) {
  return +e;
}
function Je(e) {
  return Math.floor(+e / 1e3);
}
var J, mt;
Ir({
  dateTime: "%x, %X",
  date: "%-m/%-d/%Y",
  time: "%-I:%M:%S %p",
  periods: ["AM", "PM"],
  days: ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"],
  shortDays: ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"],
  months: ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"],
  shortMonths: ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]
});
function Ir(e) {
  return J = In(e), mt = J.format, J.parse, J.utcFormat, J.utcParse, J;
}
function Ar(e) {
  return new Date(e);
}
function Pr(e) {
  return e instanceof Date ? +e : +/* @__PURE__ */ new Date(+e);
}
function dt(e, t, n, r, o, i, s, a, c, h) {
  var m = en(), d = m.invert, g = m.domain, T = h(".%L"), M = h(":%S"), U = h("%I:%M"), I = h("%I %p"), k = h("%a %d"), Q = h("%b %d"), A = h("%B"), P = h("%Y");
  function G(f) {
    return (c(f) < f ? T : a(f) < f ? M : s(f) < f ? U : i(f) < f ? I : r(f) < f ? o(f) < f ? k : Q : n(f) < f ? A : P)(f);
  }
  return m.invert = function(f) {
    return new Date(d(f));
  }, m.domain = function(f) {
    return arguments.length ? g(Array.from(f, Pr)) : g().map(Ar);
  }, m.ticks = function(f) {
    var C = g();
    return e(C[0], C[C.length - 1], f ?? 10);
  }, m.tickFormat = function(f, C) {
    return C == null ? G : h(C);
  }, m.nice = function(f) {
    var C = g();
    return (!f || typeof f.range != "function") && (f = t(C[0], C[C.length - 1], f ?? 10)), f ? g(pn(C, f)) : m;
  }, m.copy = function() {
    return tn(m, dt(e, t, n, r, o, i, s, a, c, h));
  }, m;
}
function Ai() {
  return ut.apply(dt(Ln, Hn, _, He, de, ue, Re, Fe, te, mt).domain([new Date(2e3, 0, 1), new Date(2e3, 0, 2)]), arguments);
}
const Pi = Math.abs, ki = Math.atan2, Oi = Math.cos, Wi = Math.max, $i = Math.min, Yi = Math.sin, _i = Math.sqrt, Bi = 1e-12, Ie = Math.PI, Ke = Ie / 2, Vi = 2 * Ie;
function zi(e) {
  return e > 1 ? 0 : e < -1 ? Ie : Math.acos(e);
}
function qi(e) {
  return e >= 1 ? Ke : e <= -1 ? -Ke : Math.asin(e);
}
function yt(e) {
  this._context = e;
}
yt.prototype = {
  areaStart: function() {
    this._line = 0;
  },
  areaEnd: function() {
    this._line = NaN;
  },
  lineStart: function() {
    this._point = 0;
  },
  lineEnd: function() {
    (this._line || this._line !== 0 && this._point === 1) && this._context.closePath(), this._line = 1 - this._line;
  },
  point: function(e, t) {
    switch (e = +e, t = +t, this._point) {
      case 0:
        this._point = 1, this._line ? this._context.lineTo(e, t) : this._context.moveTo(e, t);
        break;
      case 1:
        this._point = 2;
      // falls through
      default:
        this._context.lineTo(e, t);
        break;
    }
  }
};
function Gi(e) {
  return new yt(e);
}
function et() {
}
function pt(e) {
  this._context = e;
}
pt.prototype = {
  areaStart: et,
  areaEnd: et,
  lineStart: function() {
    this._point = 0;
  },
  lineEnd: function() {
    this._point && this._context.closePath();
  },
  point: function(e, t) {
    e = +e, t = +t, this._point ? this._context.lineTo(e, t) : (this._point = 1, this._context.moveTo(e, t));
  }
};
function Ni(e) {
  return new pt(e);
}
function kr(e, t) {
  if (e.match(/^[a-z]+:\/\//i))
    return e;
  if (e.match(/^\/\//))
    return window.location.protocol + e;
  if (e.match(/^[a-z]+:/i))
    return e;
  const n = document.implementation.createHTMLDocument(), r = n.createElement("base"), o = n.createElement("a");
  return n.head.appendChild(r), n.body.appendChild(o), t && (r.href = t), o.href = e, o.href;
}
const Or = /* @__PURE__ */ (() => {
  let e = 0;
  const t = () => (
    // eslint-disable-next-line no-bitwise
    `0000${(Math.random() * 36 ** 4 << 0).toString(36)}`.slice(-4)
  );
  return () => (e += 1, `u${t()}${e}`);
})();
function q(e) {
  const t = [];
  for (let n = 0, r = e.length; n < r; n++)
    t.push(e[n]);
  return t;
}
let K = null;
function wt(e = {}) {
  return K || (e.includeStyleProperties ? (K = e.includeStyleProperties, K) : (K = q(window.getComputedStyle(document.documentElement)), K));
}
function ge(e, t) {
  const r = (e.ownerDocument.defaultView || window).getComputedStyle(e).getPropertyValue(t);
  return r ? parseFloat(r.replace("px", "")) : 0;
}
function Wr(e) {
  const t = ge(e, "border-left-width"), n = ge(e, "border-right-width");
  return e.clientWidth + t + n;
}
function $r(e) {
  const t = ge(e, "border-top-width"), n = ge(e, "border-bottom-width");
  return e.clientHeight + t + n;
}
function Tt(e, t = {}) {
  const n = t.width || Wr(e), r = t.height || $r(e);
  return { width: n, height: r };
}
function Yr() {
  let e, t;
  try {
    t = process;
  } catch {
  }
  const n = t && t.env ? t.env.devicePixelRatio : null;
  return n && (e = parseInt(n, 10), Number.isNaN(e) && (e = 1)), e || window.devicePixelRatio || 1;
}
const L = 16384;
function _r(e) {
  (e.width > L || e.height > L) && (e.width > L && e.height > L ? e.width > e.height ? (e.height *= L / e.width, e.width = L) : (e.width *= L / e.height, e.height = L) : e.width > L ? (e.height *= L / e.width, e.width = L) : (e.width *= L / e.height, e.height = L));
}
function me(e) {
  return new Promise((t, n) => {
    const r = new Image();
    r.onload = () => {
      r.decode().then(() => {
        requestAnimationFrame(() => t(r));
      });
    }, r.onerror = n, r.crossOrigin = "anonymous", r.decoding = "async", r.src = e;
  });
}
async function Br(e) {
  return Promise.resolve().then(() => new XMLSerializer().serializeToString(e)).then(encodeURIComponent).then((t) => `data:image/svg+xml;charset=utf-8,${t}`);
}
async function Vr(e, t, n) {
  const r = "http://www.w3.org/2000/svg", o = document.createElementNS(r, "svg"), i = document.createElementNS(r, "foreignObject");
  return o.setAttribute("width", `${t}`), o.setAttribute("height", `${n}`), o.setAttribute("viewBox", `0 0 ${t} ${n}`), i.setAttribute("width", "100%"), i.setAttribute("height", "100%"), i.setAttribute("x", "0"), i.setAttribute("y", "0"), i.setAttribute("externalResourcesRequired", "true"), o.appendChild(i), i.appendChild(e), Br(o);
}
const v = (e, t) => {
  if (e instanceof t)
    return !0;
  const n = Object.getPrototypeOf(e);
  return n === null ? !1 : n.constructor.name === t.name || v(n, t);
};
function zr(e) {
  const t = e.getPropertyValue("content");
  return `${e.cssText} content: '${t.replace(/'|"/g, "")}';`;
}
function qr(e, t) {
  return wt(t).map((n) => {
    const r = e.getPropertyValue(n), o = e.getPropertyPriority(n);
    return `${n}: ${r}${o ? " !important" : ""};`;
  }).join(" ");
}
function Gr(e, t, n, r) {
  const o = `.${e}:${t}`, i = n.cssText ? zr(n) : qr(n, r);
  return document.createTextNode(`${o}{${i}}`);
}
function tt(e, t, n, r) {
  const o = window.getComputedStyle(e, n), i = o.getPropertyValue("content");
  if (i === "" || i === "none")
    return;
  const s = Or();
  try {
    t.className = `${t.className} ${s}`;
  } catch {
    return;
  }
  const a = document.createElement("style");
  a.appendChild(Gr(s, n, o, r)), t.appendChild(a);
}
function Nr(e, t, n) {
  tt(e, t, ":before", n), tt(e, t, ":after", n);
}
const nt = "application/font-woff", rt = "image/jpeg", jr = {
  woff: nt,
  woff2: nt,
  ttf: "application/font-truetype",
  eot: "application/vnd.ms-fontobject",
  png: "image/png",
  jpg: rt,
  jpeg: rt,
  gif: "image/gif",
  tiff: "image/tiff",
  svg: "image/svg+xml",
  webp: "image/webp"
};
function Zr(e) {
  const t = /\.([^./]*?)$/g.exec(e);
  return t ? t[1] : "";
}
function Ae(e) {
  const t = Zr(e).toLowerCase();
  return jr[t] || "";
}
function Xr(e) {
  return e.split(/,/)[1];
}
function Ee(e) {
  return e.search(/^(data:)/) !== -1;
}
function Qr(e, t) {
  return `data:${t};base64,${e}`;
}
async function St(e, t, n) {
  const r = await fetch(e, t);
  if (r.status === 404)
    throw new Error(`Resource "${r.url}" not found`);
  const o = await r.blob();
  return new Promise((i, s) => {
    const a = new FileReader();
    a.onerror = s, a.onloadend = () => {
      try {
        i(n({ res: r, result: a.result }));
      } catch (c) {
        s(c);
      }
    }, a.readAsDataURL(o);
  });
}
const be = {};
function Jr(e, t, n) {
  let r = e.replace(/\?.*/, "");
  return n && (r = e), /ttf|otf|eot|woff2?/i.test(r) && (r = r.replace(/.*\//, "")), t ? `[${t}]${r}` : r;
}
async function Pe(e, t, n) {
  const r = Jr(e, t, n.includeQueryParams);
  if (be[r] != null)
    return be[r];
  n.cacheBust && (e += (/\?/.test(e) ? "&" : "?") + (/* @__PURE__ */ new Date()).getTime());
  let o;
  try {
    const i = await St(e, n.fetchRequestInit, ({ res: s, result: a }) => (t || (t = s.headers.get("Content-Type") || ""), Xr(a)));
    o = Qr(i, t);
  } catch (i) {
    o = n.imagePlaceholder || "";
    let s = `Failed to fetch resource: ${e}`;
    i && (s = typeof i == "string" ? i : i.message), s && console.warn(s);
  }
  return be[r] = o, o;
}
async function Kr(e) {
  const t = e.toDataURL();
  return t === "data:," ? e.cloneNode(!1) : me(t);
}
async function ei(e, t) {
  if (e.currentSrc) {
    const i = document.createElement("canvas"), s = i.getContext("2d");
    i.width = e.clientWidth, i.height = e.clientHeight, s == null || s.drawImage(e, 0, 0, i.width, i.height);
    const a = i.toDataURL();
    return me(a);
  }
  const n = e.poster, r = Ae(n), o = await Pe(n, r, t);
  return me(o);
}
async function ti(e, t) {
  var n;
  try {
    if (!((n = e == null ? void 0 : e.contentDocument) === null || n === void 0) && n.body)
      return await ye(e.contentDocument.body, t, !0);
  } catch {
  }
  return e.cloneNode(!1);
}
async function ni(e, t) {
  return v(e, HTMLCanvasElement) ? Kr(e) : v(e, HTMLVideoElement) ? ei(e, t) : v(e, HTMLIFrameElement) ? ti(e, t) : e.cloneNode(Ct(e));
}
const ri = (e) => e.tagName != null && e.tagName.toUpperCase() === "SLOT", Ct = (e) => e.tagName != null && e.tagName.toUpperCase() === "SVG";
async function ii(e, t, n) {
  var r, o;
  if (Ct(t))
    return t;
  let i = [];
  return ri(e) && e.assignedNodes ? i = q(e.assignedNodes()) : v(e, HTMLIFrameElement) && (!((r = e.contentDocument) === null || r === void 0) && r.body) ? i = q(e.contentDocument.body.childNodes) : i = q(((o = e.shadowRoot) !== null && o !== void 0 ? o : e).childNodes), i.length === 0 || v(e, HTMLVideoElement) || await i.reduce((s, a) => s.then(() => ye(a, n)).then((c) => {
    c && t.appendChild(c);
  }), Promise.resolve()), t;
}
function oi(e, t, n) {
  const r = t.style;
  if (!r)
    return;
  const o = window.getComputedStyle(e);
  o.cssText ? (r.cssText = o.cssText, r.transformOrigin = o.transformOrigin) : wt(n).forEach((i) => {
    let s = o.getPropertyValue(i);
    i === "font-size" && s.endsWith("px") && (s = `${Math.floor(parseFloat(s.substring(0, s.length - 2))) - 0.1}px`), v(e, HTMLIFrameElement) && i === "display" && s === "inline" && (s = "block"), i === "d" && t.getAttribute("d") && (s = `path(${t.getAttribute("d")})`), r.setProperty(i, s, o.getPropertyPriority(i));
  });
}
function si(e, t) {
  v(e, HTMLTextAreaElement) && (t.innerHTML = e.value), v(e, HTMLInputElement) && t.setAttribute("value", e.value);
}
function ai(e, t) {
  if (v(e, HTMLSelectElement)) {
    const n = t, r = Array.from(n.children).find((o) => e.value === o.getAttribute("value"));
    r && r.setAttribute("selected", "");
  }
}
function ui(e, t, n) {
  return v(t, Element) && (oi(e, t, n), Nr(e, t, n), si(e, t), ai(e, t)), t;
}
async function ci(e, t) {
  const n = e.querySelectorAll ? e.querySelectorAll("use") : [];
  if (n.length === 0)
    return e;
  const r = {};
  for (let i = 0; i < n.length; i++) {
    const a = n[i].getAttribute("xlink:href");
    if (a) {
      const c = e.querySelector(a), h = document.querySelector(a);
      !c && h && !r[a] && (r[a] = await ye(h, t, !0));
    }
  }
  const o = Object.values(r);
  if (o.length) {
    const i = "http://www.w3.org/1999/xhtml", s = document.createElementNS(i, "svg");
    s.setAttribute("xmlns", i), s.style.position = "absolute", s.style.width = "0", s.style.height = "0", s.style.overflow = "hidden", s.style.display = "none";
    const a = document.createElementNS(i, "defs");
    s.appendChild(a);
    for (let c = 0; c < o.length; c++)
      a.appendChild(o[c]);
    e.appendChild(s);
  }
  return e;
}
async function ye(e, t, n) {
  return !n && t.filter && !t.filter(e) ? null : Promise.resolve(e).then((r) => ni(r, t)).then((r) => ii(e, r, t)).then((r) => ui(e, r, t)).then((r) => ci(r, t));
}
const Mt = /url\((['"]?)([^'"]+?)\1\)/g, li = /url\([^)]+\)\s*format\((["']?)([^"']+)\1\)/g, fi = /src:\s*(?:url\([^)]+\)\s*format\([^)]+\)[,;]\s*)+/g;
function hi(e) {
  const t = e.replace(/([.*+?^${}()|\[\]\/\\])/g, "\\$1");
  return new RegExp(`(url\\(['"]?)(${t})(['"]?\\))`, "g");
}
function gi(e) {
  const t = [];
  return e.replace(Mt, (n, r, o) => (t.push(o), n)), t.filter((n) => !Ee(n));
}
async function mi(e, t, n, r, o) {
  try {
    const i = n ? kr(t, n) : t, s = Ae(t);
    let a;
    return o || (a = await Pe(i, s, r)), e.replace(hi(t), `$1${a}$3`);
  } catch {
  }
  return e;
}
function di(e, { preferredFontFormat: t }) {
  return t ? e.replace(fi, (n) => {
    for (; ; ) {
      const [r, , o] = li.exec(n) || [];
      if (!o)
        return "";
      if (o === t)
        return `src: ${r};`;
    }
  }) : e;
}
function xt(e) {
  return e.search(Mt) !== -1;
}
async function bt(e, t, n) {
  if (!xt(e))
    return e;
  const r = di(e, n);
  return gi(r).reduce((i, s) => i.then((a) => mi(a, s, t, n)), Promise.resolve(r));
}
async function ee(e, t, n) {
  var r;
  const o = (r = t.style) === null || r === void 0 ? void 0 : r.getPropertyValue(e);
  if (o) {
    const i = await bt(o, null, n);
    return t.style.setProperty(e, i, t.style.getPropertyPriority(e)), !0;
  }
  return !1;
}
async function yi(e, t) {
  await ee("background", e, t) || await ee("background-image", e, t), await ee("mask", e, t) || await ee("-webkit-mask", e, t) || await ee("mask-image", e, t) || await ee("-webkit-mask-image", e, t);
}
async function pi(e, t) {
  const n = v(e, HTMLImageElement);
  if (!(n && !Ee(e.src)) && !(v(e, SVGImageElement) && !Ee(e.href.baseVal)))
    return;
  const r = n ? e.src : e.href.baseVal, o = await Pe(r, Ae(r), t);
  await new Promise((i, s) => {
    e.onload = i, e.onerror = t.onImageErrorHandler ? (...c) => {
      try {
        i(t.onImageErrorHandler(...c));
      } catch (h) {
        s(h);
      }
    } : s;
    const a = e;
    a.decode && (a.decode = i), a.loading === "lazy" && (a.loading = "eager"), n ? (e.srcset = "", e.src = o) : e.href.baseVal = o;
  });
}
async function wi(e, t) {
  const r = q(e.childNodes).map((o) => Dt(o, t));
  await Promise.all(r).then(() => e);
}
async function Dt(e, t) {
  v(e, Element) && (await yi(e, t), await pi(e, t), await wi(e, t));
}
function Ti(e, t) {
  const { style: n } = e;
  t.backgroundColor && (n.backgroundColor = t.backgroundColor), t.width && (n.width = `${t.width}px`), t.height && (n.height = `${t.height}px`);
  const r = t.style;
  return r != null && Object.keys(r).forEach((o) => {
    n[o] = r[o];
  }), e;
}
const it = {};
async function ot(e) {
  let t = it[e];
  if (t != null)
    return t;
  const r = await (await fetch(e)).text();
  return t = { url: e, cssText: r }, it[e] = t, t;
}
async function st(e, t) {
  let n = e.cssText;
  const r = /url\(["']?([^"')]+)["']?\)/g, i = (n.match(/url\([^)]+\)/g) || []).map(async (s) => {
    let a = s.replace(r, "$1");
    return a.startsWith("https://") || (a = new URL(a, e.url).href), St(a, t.fetchRequestInit, ({ result: c }) => (n = n.replace(s, `url(${c})`), [s, c]));
  });
  return Promise.all(i).then(() => n);
}
function at(e) {
  if (e == null)
    return [];
  const t = [], n = /(\/\*[\s\S]*?\*\/)/gi;
  let r = e.replace(n, "");
  const o = new RegExp("((@.*?keyframes [\\s\\S]*?){([\\s\\S]*?}\\s*?)})", "gi");
  for (; ; ) {
    const c = o.exec(r);
    if (c === null)
      break;
    t.push(c[0]);
  }
  r = r.replace(o, "");
  const i = /@import[\s\S]*?url\([^)]*\)[\s\S]*?;/gi, s = "((\\s*?(?:\\/\\*[\\s\\S]*?\\*\\/)?\\s*?@media[\\s\\S]*?){([\\s\\S]*?)}\\s*?})|(([\\s\\S]*?){([\\s\\S]*?)})", a = new RegExp(s, "gi");
  for (; ; ) {
    let c = i.exec(r);
    if (c === null) {
      if (c = a.exec(r), c === null)
        break;
      i.lastIndex = a.lastIndex;
    } else
      a.lastIndex = i.lastIndex;
    t.push(c[0]);
  }
  return t;
}
async function Si(e, t) {
  const n = [], r = [];
  return e.forEach((o) => {
    if ("cssRules" in o)
      try {
        q(o.cssRules || []).forEach((i, s) => {
          if (i.type === CSSRule.IMPORT_RULE) {
            let a = s + 1;
            const c = i.href, h = ot(c).then((m) => st(m, t)).then((m) => at(m).forEach((d) => {
              try {
                o.insertRule(d, d.startsWith("@import") ? a += 1 : o.cssRules.length);
              } catch (g) {
                console.error("Error inserting rule from remote css", {
                  rule: d,
                  error: g
                });
              }
            })).catch((m) => {
              console.error("Error loading remote css", m.toString());
            });
            r.push(h);
          }
        });
      } catch (i) {
        const s = e.find((a) => a.href == null) || document.styleSheets[0];
        o.href != null && r.push(ot(o.href).then((a) => st(a, t)).then((a) => at(a).forEach((c) => {
          s.insertRule(c, s.cssRules.length);
        })).catch((a) => {
          console.error("Error loading remote stylesheet", a);
        })), console.error("Error inlining remote css file", i);
      }
  }), Promise.all(r).then(() => (e.forEach((o) => {
    if ("cssRules" in o)
      try {
        q(o.cssRules || []).forEach((i) => {
          n.push(i);
        });
      } catch (i) {
        console.error(`Error while reading CSS rules from ${o.href}`, i);
      }
  }), n));
}
function Ci(e) {
  return e.filter((t) => t.type === CSSRule.FONT_FACE_RULE).filter((t) => xt(t.style.getPropertyValue("src")));
}
async function Mi(e, t) {
  if (e.ownerDocument == null)
    throw new Error("Provided element is not within a Document");
  const n = q(e.ownerDocument.styleSheets), r = await Si(n, t);
  return Ci(r);
}
function Et(e) {
  return e.trim().replace(/["']/g, "");
}
function xi(e) {
  const t = /* @__PURE__ */ new Set();
  function n(r) {
    (r.style.fontFamily || getComputedStyle(r).fontFamily).split(",").forEach((i) => {
      t.add(Et(i));
    }), Array.from(r.children).forEach((i) => {
      i instanceof HTMLElement && n(i);
    });
  }
  return n(e), t;
}
async function bi(e, t) {
  const n = await Mi(e, t), r = xi(e);
  return (await Promise.all(n.filter((i) => r.has(Et(i.style.fontFamily))).map((i) => {
    const s = i.parentStyleSheet ? i.parentStyleSheet.href : null;
    return bt(i.cssText, s, t);
  }))).join(`
`);
}
async function Di(e, t) {
  const n = t.fontEmbedCSS != null ? t.fontEmbedCSS : t.skipFonts ? null : await bi(e, t);
  if (n) {
    const r = document.createElement("style"), o = document.createTextNode(n);
    r.appendChild(o), e.firstChild ? e.insertBefore(r, e.firstChild) : e.appendChild(r);
  }
}
async function Ei(e, t = {}) {
  const { width: n, height: r } = Tt(e, t), o = await ye(e, t, !0);
  return await Di(o, t), await Dt(o, t), Ti(o, t), await Vr(o, n, r);
}
async function vt(e, t = {}) {
  const { width: n, height: r } = Tt(e, t), o = await Ei(e, t), i = await me(o), s = document.createElement("canvas"), a = s.getContext("2d"), c = t.pixelRatio || Yr(), h = t.canvasWidth || n, m = t.canvasHeight || r;
  return s.width = h * c, s.height = m * c, t.skipAutoScale || _r(s), s.style.width = `${h}`, s.style.height = `${m}`, t.backgroundColor && (a.fillStyle = t.backgroundColor, a.fillRect(0, 0, s.width, s.height)), a.drawImage(i, 0, 0, s.width, s.height), s;
}
async function vi(e, t = {}) {
  return (await vt(e, t)).toDataURL();
}
async function Ui(e, t = {}) {
  return (await vt(e, t)).toDataURL("image/jpeg", t.quality || 1);
}
class Ut {
  constructor(t, n) {
    this.model = t, this.services = n, this.init();
  }
  init() {
  }
  update() {
  }
  // Used to pass down information to the components
  setModel(t) {
    this.model = t;
  }
  // Used to pass down services to the components
  setServices(t) {
    this.services = t;
  }
}
const Fi = "DONT_STYLE_ME_css_styles_verifier";
class ve extends Ut {
  // initialized in initializeID() called by init()
  constructor(t, n) {
    super(t, n);
  }
  static getHTMLElementSize(t) {
    return {
      width: t.clientWidth,
      height: t.clientHeight
    };
  }
  static getSVGElementSize(t, n = {
    useAttrs: !1,
    useClientDimensions: !1,
    useBBox: !1,
    useBoundingRect: !1
  }) {
    t.attr || (t = D(t));
    const r = {
      width: 0,
      height: 0
    }, o = (g) => {
      g && Object.keys(r).forEach((T) => {
        if (g[T]) {
          const M = g[T], U = parseFloat(M);
          M && U > r[T] && ("" + M).indexOf("%") === -1 && (r[T] = U);
        }
      });
    }, i = {
      width: t.attr("width"),
      height: t.attr("height")
    }, s = t.node();
    let a, c, h, m;
    try {
      typeof s.getBBox == "function" && (a = s.getBBox(), c = {
        width: a.width,
        height: a.height
      });
    } catch (g) {
      console.error(g);
    }
    try {
      typeof (s == null ? void 0 : s.getBoundingClientRect) == "function" && (h = s.getBoundingClientRect(), m = {
        width: h.width,
        height: h.height
      });
    } catch (g) {
      console.error(g);
    }
    let d;
    if (s instanceof SVGSVGElement && (d = {
      width: s.clientWidth,
      height: s.clientHeight
    }), n) {
      if (n.useAttrs && (o(i), r.width > 0 && r.height > 0))
        return r;
      if (n.useClientDimensions && (o(d), r.width > 0 && r.height > 0))
        return d;
      if (n.useBBox && (o(c), r.width > 0 && r.height > 0))
        return c;
      if (n.useBoundingRect && (o(m), r.width > 0 && r.height > 0))
        return m;
    }
    try {
      const g = {
        width: ie(t.node(), "width", "baseVal", "value"),
        height: ie(t.node(), "height", "baseVal", "value")
      };
      o(g);
    } catch {
      o(d), o(c), o(i);
    }
    return r;
  }
  static appendOrSelect(t, n) {
    const r = t.select(`${n}`);
    if (r.empty()) {
      let o = n.split("#"), i, s;
      return o.length === 2 ? (i = o[0], o = o[1].split("."), s = o[0]) : (o = n.split("."), i = o[0]), t.append(i).attr("id", s).attr("class", o.slice(1).join(" "));
    }
    return r;
  }
  init() {
    this.initializeID(), this.styleHolderElement(), this.addMainContainer(), this.model.getOptions().resizable && this.addResizeListener(), this.addHolderListeners(), this.handleFullscreenChange();
  }
  getChartID() {
    return this.chartID;
  }
  getElementOffset(t, n = !1) {
    const r = { left: 0, top: 0 }, o = t.getBoundingClientRect(), i = n ? { left: 0, top: 0 } : this.getHolder().getBoundingClientRect();
    try {
      r.left = o.left - i.left, r.top = o.top - i.top;
    } catch (s) {
      console.error(s);
    }
    return r;
  }
  generateElementIDString(t) {
    return `chart-${this.chartID}-${t}`;
  }
  initializeID() {
    this.chartID = Math.floor((1 + Math.random()) * 281474976710656).toString(16);
  }
  addMainContainer() {
    const t = this.model.getOptions(), n = ie(t, "style", "prefix"), r = D(this.getHolder()).append("div").classed(`${$e}--${n}--chart-wrapper`, !0).attr("id", `chart-${this.getChartID()}`).style("height", "100%").style("width", "100%");
    r.append("g").attr("class", Fi), this.mainContainer = r.node();
  }
  update() {
    this.styleHolderElement();
  }
  styleHolderElement() {
    const t = this.getHolder(), { width: n, height: r, theme: o } = this.model.getOptions();
    n !== this.width && (t.style.width = n, this.width = n), r !== this.height && (t.style.height = r, this.height = r), D(this.getHolder()).classed(`${$e}--chart-holder`, !0).attr("data-carbon-theme", o);
  }
  getHolder() {
    return this.model.get("holder");
  }
  exportToJPG() {
    const t = this, n = this.model.getOptions(), r = this.getHolder(), o = D(r);
    o.classed("filled", !0), Ui(this.getMainContainer(), {
      quality: 1,
      // Remove toolbar
      filter: (i) => !(i.classList && i.classList.contains("cds--cc--toolbar"))
    }).then(function(i) {
      var c;
      let s = "myChart";
      const a = ie(n, "fileDownload", "fileName");
      typeof a == "function" ? s = a("jpg") : typeof a == "string" && (s = a), (c = t.services.files) == null || c.downloadImage(i, `${s}.jpg`), o.classed("filled", !1);
    });
  }
  exportToPNG() {
    const t = this, n = this.model.getOptions(), r = this.getHolder(), o = D(r);
    o.classed("filled", !0), vi(this.getMainContainer(), {
      quality: 1,
      // Remove toolbar
      filter: (i) => !(i.classList && i.classList.contains("cds--cc--toolbar"))
    }).then(function(i) {
      var c;
      let s = "myChart";
      const a = ie(n, "fileDownload", "fileName");
      typeof a == "function" ? s = a("png") : typeof a == "string" && (s = a), (c = t.services.files) == null || c.downloadImage(i, `${s}.png`), o.classed("filled", !1);
    }).catch(function(i) {
      console.error("oops, something went wrong!", i);
    });
  }
  isFullScreenMode() {
    return D(this.getHolder()).classed("fullscreen");
  }
  toggleFullscreen() {
    const t = this.getHolder();
    D(t).classed("fullscreen") && (document.fullscreenElement || document.webkitFullscreenElement || document.mozFullScreenElement || document.msFullscreenElement) ? document.exitFullscreen ? document.exitFullscreen() : document.webkitExitFullscreen ? document.webkitExitFullscreen() : document.mozCancelFullScreen ? document.mozCancelFullScreen() : document.msExitFullscreen && document.msExitFullscreen() : t.requestFullscreen ? t.requestFullscreen() : t.webkitRequestFullscreen ? t.webkitRequestFullscreen() : t.mozRequestFullScreen ? t.mozRequestFullScreen() : t.msRequestFullscreen && t.msRequestFullscreen();
  }
  handleFullscreenChange() {
    document.addEventListener("fullscreenchange", () => {
      const t = D(this.getHolder()), n = t.classed("fullscreen");
      t.classed("fullscreen", !n);
    });
  }
  setSVGMaxHeight() {
    if (!this.model.getOptions().height) {
      const { height: t } = ve.getSVGElementSize(
        D(this.mainContainer),
        {
          useBBox: !0
        }
      ), n = D(this.mainContainer).attr("class"), r = D(this.mainContainer).selectAll(`.${n} > svg`);
      let o = 0;
      r.nodes().forEach(function(i) {
        o += Number(
          ve.getSVGElementSize(D(i), {
            useBBox: !0
          }).height
        );
      }), o <= t ? D(this.mainContainer).attr("height", o) : D(this.mainContainer).attr("height", "100%");
    }
  }
  getMainContainer() {
    return this.mainContainer;
  }
  addHolderListeners() {
    const t = this.getHolder();
    t && D(t).on("mouseover", () => {
      var n;
      (n = this.services.events) == null || n.dispatchEvent(pe.Chart.MOUSEOVER);
    }).on("mouseout", () => {
      var n;
      (n = this.services.events) == null || n.dispatchEvent(pe.Chart.MOUSEOUT);
    });
  }
  addResizeListener() {
    const t = this.getHolder();
    if (!t)
      return;
    let n = t.clientWidth, r = t.clientHeight;
    const o = mn(() => {
      var s;
      t && (Math.abs(n - t.clientWidth) > 1 || Math.abs(r - t.clientHeight) > 1) && (n = t.clientWidth, r = t.clientHeight, (s = this.services.events) == null || s.dispatchEvent(pe.Chart.RESIZE));
    }, 12.5);
    new ResizeObserver(o).observe(t);
  }
}
class Ft extends Ut {
  static appendOrUpdateLinearGradient(t) {
    let n = t.svg.select(`defs linearGradient#${t.id}`);
    n.empty() && (n = t.svg.append("defs").append("linearGradient").attr("id", t.id).attr("x1", t.x1).attr("x2", t.x2).attr("y1", t.y1).attr("y2", t.y2)), n.selectAll("stop").remove(), n.selectAll("stop").data(t.stops).enter().append("stop").attr("offset", (r) => r.offset).style("stop-color", (r) => r.color).style("stop-opacity", (r) => r.opacity);
  }
  static getOffsetRatio(t) {
    return (Math.abs(t[1]) * 100 / Math.abs(t[0] - t[1])).toFixed(2) + "%";
  }
  static getStops(t, n) {
    const r = t[0] < 0 && t[1] > 0;
    let o = [
      {
        offset: "0%",
        color: n,
        opacity: "0.6"
      },
      {
        offset: "80%",
        color: n,
        opacity: "0"
      }
    ];
    return r && (o = [
      {
        offset: "0%",
        color: n,
        opacity: "0.6"
      },
      {
        offset: Ft.getOffsetRatio(t),
        color: n,
        opacity: "0"
      },
      {
        offset: "100%",
        color: n,
        opacity: "0.6"
      }
    ]), o;
  }
}
function ji(e) {
  const t = Ri(Li(e), 360);
  return z(t, [0, 10]) || z(t, [350, 0]) ? {
    textAnchor: V.START,
    dominantBaseline: B.MIDDLE
  } : z(t, [10, 80]) ? {
    textAnchor: V.START,
    dominantBaseline: B.HANGING
  } : z(t, [80, 100]) ? {
    textAnchor: V.MIDDLE,
    dominantBaseline: B.HANGING
  } : z(t, [100, 170]) ? {
    textAnchor: V.END,
    dominantBaseline: B.HANGING
  } : z(t, [170, 190]) ? {
    textAnchor: V.END,
    dominantBaseline: B.MIDDLE
  } : z(t, [190, 260]) ? {
    textAnchor: V.END,
    dominantBaseline: B.BASELINE
  } : z(t, [260, 280]) ? {
    textAnchor: V.MIDDLE,
    dominantBaseline: B.BASELINE
  } : {
    textAnchor: V.START,
    dominantBaseline: B.BASELINE
  };
}
function Ri(e, t) {
  return (e % t + t) % t;
}
function z(e, [t, n]) {
  return e >= t && e <= n;
}
function Li(e) {
  return e * (180 / Math.PI);
}
function Zi(e) {
  return e * (Math.PI / 180);
}
function Xi(e, t, n = { x: 0, y: 0 }) {
  const r = t * Math.cos(e) + n.x, o = t * Math.sin(e) + n.y;
  return { x: isNaN(r) ? 0 : r, y: isNaN(o) ? 0 : o };
}
function Qi(e, t) {
  return t * Math.sin(e - Math.PI / 2);
}
export {
  et as A,
  ve as D,
  Ft as G,
  Ut as S,
  Qi as a,
  Li as b,
  Vi as c,
  Zi as d,
  Bi as e,
  Oi as f,
  Yi as g,
  Ke as h,
  Pi as i,
  Ie as j,
  _i as k,
  ki as l,
  $i as m,
  qi as n,
  zi as o,
  Xi as p,
  Wi as q,
  ji as r,
  D as s,
  _e as t,
  Gi as u,
  Ai as v,
  mn as w,
  yn as x,
  Ni as y,
  pn as z
};
